man/man1/lpr.1
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\" @(#)lpr.1 1.1 92/07/30 SMI; from UCB 4.3 BSD 6.1
.\"
.TH LPR 1 "28 August 1989"
.SH NAME
lpr \- send a job to the printer
.SH SYNOPSIS
.B lpr
[
.BI \-P printer
]
[
.BI \-# copies
]
[
.BI \-C class
]
[
.BI \-J job
]
.if n .ti +5n
[
.BI \-T title
]
[
.B \-i
[
.I indent
] ] [
.BI \-1234 font
]
.if t .ti +.5i
[
.BI \-w cols
]
.if n .ti +5n
[
.B \-r
] [
.B \-m
] [
.B \-h
] [
.B \-s
] [
.BI \- filter-option
]
.if n .ti +5n
[
.I filename
\&.\|.\|.
]
.SH DESCRIPTION
.IX  "lpr command"  ""  "\fLlpr\fP \(em print files"
.IX  "print files" "" "print files \(em \fLlpr\fP"
.IX  file  print  ""  "print \(em \fLlpr\fP"
.LP
.B lpr
creates a printer job in a spooling area for subsequent printing as
facilities become available.
Each printer job consists of a control file and one or more 
data files.  The data files are
copies of
(or, with
.B \-s ,
symbolic links to) each
.I filename
you specify.
The spool area is managed by the line printer daemon,
.BR lpd (8).
Jobs that specify a printer on a remote machine are
forwarded by
.BR lpd .
.LP
.B lpr
reads from the standard input if no files are specified.
.SH OPTIONS
.LP
.TP 15
.BI \-P printer
Send output to the named
.IR printer .
Otherwise send output to the printer named in the
.SB PRINTER
environment variable, or to the default printer,
.BR lp .
.TP
.BI \-# copies
Produce the number of
.I copies
indicated for each named file.  For example:
.RS
.IP
.B "example% lpr \-#3 index.c lookup.c"
.RE
.IP
produces three copies of
.BR index.c ,
followed by three copies of
.BR lookup.c .
On the other hand,
.RS
.IP
.B "example% cat index.c lookup.c | lpr \-#3"
.RE
.IP
generates three copies of the concatenation of the files.
.TP
.BI \-C class
Print
.I class
as the job classification on the
burst page.
For example,
.RS
.IP
.B example% lpr \-C Operations new.index.c
.RE
.IP
replaces the system name (the name returned by
.IR hostname )
with \*(lqOperations\*(rq on the burst page, and prints the file
.IR new.index.c .
.TP
.BI \-J job
Print
.I job
as the job name on the
burst page.  Normally,
.B lpr
uses the first file's name.
.TP
.BI \-T title
Use
.I title
instead of the file name for the title
used by
.BR pr (1V).
.TP
\fB\-i\fR[ \fIindent\fR ]
Indent output
.I indent
.SM SPACE
characters.  Eight
.SM SPACE
characters is the default.
The indent is passed to the input filter.
If no input filter is present, this option
is ignored.
.PD 0
.TP
.BI \-1 " font "
.TP
.BI \-2 " font "
.TP
.BI \-3 " font "
.TP
.BI \-4 " font "
Mount the specified
.I font
on font position
.BR 1 ,
.BR 2 ,
.B 3
or
.BR 4 .
The daemon will construct a
.I .railmag
file in the spool directory that
indicates the mount by referencing
.BR /usr/lib/vfont/font .
.PD
.TP
.BI \-w cols
Use
.I cols
as the page width for
.BR pr .
.TP
.B \-r
Remove the file upon completion of spooling,
or upon completion of printing with the
.B \-s
option.
.TP
.B \-m
Send mail upon completion.
.TP
.B \-h
Suppress printing the burst page.
.TP
.B \-s
Create a symbolic link from the spool
area to the data files rather than trying to copy
them (so large files can be printed).
This means the data files should
not be modified or removed until they
have been printed.
This option can be used to avoid truncating
files larger than the maximum given in the
.B mx
capability of the
.BR printcap (5)
entry.
.B \-s
only prevents copies of local files from being made. 
Jobs from
remote hosts are copied anyway.
.B \-s
only works with named data files; if the
.B lpr
command is at the end of a pipeline, the data is copied to the
spool.
.br
.ne 10
.TP
.I filter-option
The following single letter options notify the line printer
spooler that the files are not standard text files.
The spooling daemon will use the appropriate filters to
print the data accordingly.
.ne 12
.RS
.TP
.B \-p
.PD 0
Use
.B pr
to format the files
.RB ( "lpr \-p"
is very much like
.RB ` "pr | lpr" ').
.TP
.B \-l
Print control characters and suppress page breaks.
.TP
.B \-t
The files contain
.BR troff (1)
(cat phototypesetter) binary data.
.TP
.B \-n
The files contain data from
.B ditroff
(device independent troff).
.TP
.B \-d
The files contain data from
.B tex
(\s-1DVI\s0
format from Stanford).
.TP
.B \-g
The files contain standard plot data as produced by the
.BR plot (3X)
routines (see also
.BR plot (1G)
for the filters used by the printer spooler).
.TP
.B \-v
The files contain a raster image, see
.BR rasterfile (5).
The printer must support an appropriate imaging model such as
PostScript in order to print the image.
.TP
.B \-c
The files contain data produced by
.BR cifplot .
.TP
.B \-f
Interpret the first character of each line as a standard
.SM FORTRAN
carriage control character.
.PD
.LP
If no
.I filter-option
is given (and the printer can interpret PostScript), the string
.RB ` %! '
as the first two characters of a file indicates that it
contains PostScript commands.
.LP
These filter options offer a standard user interface,
and all options may not be available for, nor applicable to,
all printers.
.RE
.SH FILES
.PD 0
.TP 20
.B /etc/passwd
personal identification
.TP
.B /etc/printcap
printer capabilities data base
.TP
.B /usr/lib/lpd
line printer daemon
.TP
.B /var/spool/l*
directories used for spooling
.TP
.B /var/spool/l*/cf*
daemon control files
.TP
.B /var/spool/l*/df*
data files specified in
.RB ` cf '
files
.TP
.B /var/spool/l*/tf*
temporary copies of
.RB ` cf '
files
.TP
.B /usr/lib/vfont/font
.PD
.SH "SEE ALSO"
.BR lpq (1),
.BR lprm (1),
.BR plot (1G),
.BR pr (1V),
.BR screendump (1),
.BR troff (1),
.BR plot (3X),
.BR printcap (5),
.BR rasterfile (5),
.BR lpc (8),
.BR lpd (8)
.SH DIAGNOSTICS
.TP
.B lpr: copy file is too large
A file is determined to be too \*(lqlarge\*(rq
to print by copying into the spool area.
.BR lpr
truncates the file, and prints as much of it as it can.
The maximum file length is specified by the
.B mx
capability of the
.BR printcap (5)
entry for the printer.  If no
.B mx
capability is specified, the default limit is 1000 Kbytes.
Use the
.B \-s
option as defined above to make a symbolic
link to the file instead of copying
it. 
.TP
.B lpr: \fIprinter\fP\|: unknown printer
The
.B printer
was not found in the
.B printcap
database.  Usually this is a typing mistake; however, it may indicate
a missing or incorrect entry in the
.B /etc/printcap
file.
.TP
.B "lpr: \fIprinter\fP\|: jobs queued, but cannot start daemon."
The connection to
.B lpd
on the local machine failed.
This usually means the printer server started at
boot time has died or is hung.  Check the local socket
.B /dev/printer
to be sure it still exists (if it does
not
exist,
there is no
.B lpd
process running).
.br
.ne 9
.TP
.B lpr: \fIprinter\fP\|: printer queue is disabled
This means the queue was turned off with
.RS
.IP
.BI "example% /usr/etc/lpc disable " printer
.RE
.IP
to prevent
.B lpr
from putting files in the queue.  This is normally
done by the system manager when a printer is
going to be down for a long time.  The
printer can be turned back on by a super-user with
.BR lpc .
.IP
If a connection to
.B lpd
on the local machine cannot be made
.B lpr
will say that the daemon cannot be started.
Diagnostics may be printed in the daemon log file
regarding missing spool files by
.BR lpd .
.SH BUGS
.LP
Command-line options cannot be combined into a single argument
as with some other commands.  The command:
.IP
.B lpr \-fs
.LP
is not equivalent to
.IP
.BR "lpr \-f \-s"
.LP
Placing the
.B \-s
flag first, or writing each option as a separate argument, makes a link
as expected.
.LP
.B lpr \-p
is not precisely
equivalent
to
.BR "pr | lpr" .
.B lpr \-p
puts the current date at the top of each page, rather
than the date last modified.
.LP
Fonts for
.BR troff (1)
and
\s-1T\s-1\dE\u\s+1X\s+1\(rg
reside on the printer host.
It is currently not possible to use local font libraries.
.LP
.B lpr
refuses to print a.out files and library archives.
.LP
The
.B \-s
option only avoids copying the data file to the spool directory
of the local machine.  If the printer for a job resides on a
remote machine, the data file will be copied to the remote spool
directory in all cases.


ucb/lpr/lpr.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static char sccsid[] = "%Z%%M% %I% %E% SMI"; /* from UCB 5.2 11/17/85 */
#endif not lint

/*
 *	lpr -- off line print
 *
 * Allows multiple printers and printers on remote machines by
 * using information from a printer data base.
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/param.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <pwd.h>
#include <grp.h>
#include <signal.h>
#include <ctype.h>
#include <syslog.h>
#include <strings.h>
#include <varargs.h>
#include "lp.h"

char	*tfname;		/* tmp copy of cf before linking */
char	*cfname;		/* daemon control files, linked from tf's */
char	*dfname;		/* data files */

int	nact;			/* number of jobs to act on */
int	tfd;			/* control file descriptor */
int	mailflg;		/* send mail */
int	qflag;			/* q job, but don't exec daemon */
char	format = 'f';		/* format char for printing files */
int	rflag;			/* remove files upon completion */
int	sflag;			/* symbolic link flag */
int	inchar;			/* location to increment char in file names */
int	ncopies = 1;		/* # of copies to make */
int	iflag;			/* indentation wanted */
int	indent;			/* amount to indent */
int	silent;			/* Suppress messages from lp */
int	wrtflg;			/* write to user terminal when done */
int	hdr = 1;		/* print header or not (default is yes) */
int	userid;			/* user id */
char	*person;		/* user name */
char	*title;			/* pr'ing title */
int	banner;			/* print title on banner page */
char	*bannerttl;		/* title to appear on the banner page */
char	*fonts[4];		/* troff font names */
char	*width;			/* width for versatec printing */
char	host[MAXHOSTNAMELEN + 1]; /* host name */
char	*class = host;		/* class title on header page */
char	*jobname;		/* job name on header page */
char	*name;			/* program name */
char	*printer;		/* printer name */
struct	stat statb;

int	MX;			/* maximum number of blocks to copy */
int	MC;			/* maximum number of copies allowed */
int	DU;			/* daemon user-id */
char	*SD;			/* spool directory */
char	*LO;			/* lock file name */
char	*RG;			/* restrict group */
short	SC;			/* suppress multiple copies */
char	*LP;			/* line printer */
char	*AG;			/* comma seperated list of allowed groups */

int	grp_ok;			/* 1 == member of an allowed group */

char	*getenv();
char	*getwd();
char	*getlogin();
off_t	lseek();
void	exit();
void	free();

static void	cleanup(), lprarg(), lparg();
static void	card(), copy(), chkprinter(), fatal();
static int	mktemps(), nfile(), test();
static char	*mktempfile(), *linked(), *itoa();

static int  argcnt;
static char **argvar;

/*ARGSUSED*/
main(argc, argv)
	int argc;
	char *argv[];
{
	struct passwd *pw;
	struct group *gptr;
	register char *arg, *cp;
	char buf[BUFSIZ];
	char *tty, *ttyname();
	int i, f;
	struct stat stb;
	int jobno;

	if (signal(SIGHUP, SIG_IGN) != SIG_IGN)
		(void) signal(SIGHUP, cleanup);
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		(void) signal(SIGINT, cleanup);
	if (signal(SIGQUIT, SIG_IGN) != SIG_IGN)
		(void) signal(SIGQUIT, cleanup);
	if (signal(SIGTERM, SIG_IGN) != SIG_IGN)
		(void) signal(SIGTERM, cleanup);

	cp = rindex(argv[0], '/');
	name = (cp == NULL) ? argv[0] : cp + 1;

	argcnt = argc;
	argvar = argv;

	(void)gethostname(host, sizeof host);
	openlog("lpd", 0, LOG_LPR);

	/*
	 * call lparg if we have invoked the program as "lp", 
	 * and call lprarg otherwise.
	 */
	if (strcmp(name, "lp") == 0)
		lparg();
	else
		lprarg();

	if (SC && ncopies > 1)
		fatal("multiple copies are not allowed");
	if (MC > 0 && ncopies > MC)
		fatal("only %d copies are allowed", MC);
	/*
	 * Get the identity of the person doing the lpr using the same
	 * algorithm as lprm.
	 */
	if ((person = getlogin()) == NULL || (pw = getpwnam(person)) == NULL) {
		userid = getuid();
		if ((pw = getpwuid(userid)) == NULL)
			fatal("Who are you?");
		person = pw->pw_name;
	} else
		userid = pw->pw_uid;
	/*
	 * Check for restricted group access.
	 */
	if (RG != NULL) {
		if ((gptr = getgrnam(RG)) == NULL)
			fatal("Restricted group specified incorrectly");
		if (gptr->gr_gid != getgid()) {
			while (*gptr->gr_mem != NULL) {
				if ((strcmp(person, *gptr->gr_mem)) == 0)
					break;
				gptr->gr_mem++;
			}
			if (*gptr->gr_mem == NULL)
				fatal("Not a member of the restricted group");
		}
	}
	/*
	 * Check to make sure queuing is enabled if userid is not root.
	 */
	(void) sprintf(buf, "%s/%s", SD, LO);
	if (userid && stat(buf, &stb) == 0 && (stb.st_mode & 010))
		fatal("Printer queue is disabled");
	/*
	 * Check to see if the user is a member of an allowed group.
	 */
	grp_ok = 1;	/* Assume allowed to use printer. */
	if ( (AG != NULL) && (pw->pw_uid != 0) ) {
		char *cp, *ag;
		int ngrps, i;
		struct group *gp;
		int groups[NGROUPS];

		if ((ngrps = getgroups(NGROUPS, groups)) == -1) {
			perror("getgroups");
			grp_ok = 0;
			goto LABEL;
		}
		grp_ok = 0;
		ag = malloc(strlen(AG) +1);
		(void) strcpy(ag, AG);
		cp = strtok(ag, ",");
		for (; cp != NULL; cp = strtok(NULL, ",")) {
			gp = getgrnam(cp);
			if ( gp == NULL ) {
				continue; /* Skip nonexistent groups */
			}
			for (i = 0; i < ngrps; i++) {
				if (groups[i] == gp->gr_gid) {
					grp_ok = 1;
					goto LABEL;
				}
			}
		}
	}
    LABEL:
	if ( grp_ok == 0 ) {
		char line[256];
		sprintf(line, "You are not allowed to use printer %s", printer);		fatal(line);
	}
	/*
	 * Initialize the control file.
	 */
	jobno = mktemps();
	tfd = nfile(tfname);
	(void) fchown(tfd, DU, -1);	/* owned by daemon for protection */
	card('H', host);
	card('P', person);
	if (hdr) {
		if (jobname == NULL) {
			if (argcnt == 1)
				jobname = "stdin";
			else {
				arg = rindex(argvar[1], '/');
				jobname = (arg != NULL) ? arg + 1 : argvar[1];
			}
		}
		card('J', jobname);
		card('C', class);
		if (banner)
			card('B', bannerttl);
		card('L', person);
	}
	if (iflag)
		card('I', itoa(indent));
	if (mailflg)
		card('M', person);
	if (wrtflg && !silent)
		/*
		 * write to terminal if printer is local,
		 * send mail otherwise
		 */
		if (LP != NULL && *LP != '\0') {
			if ((tty = ttyname(2)) != NULL)
				card('R', tty);
		}
		else
			card('M', person);
	if (format == 't' || format == 'n' || format == 'd')
		for (i = 0; i < 4; i++)
			if (fonts[i] != NULL)
				card('1' + i, fonts[i]);
	if (width != NULL)
		card('W', width);

	/*
	 * Read the files and spool them.
	 */
	if (argcnt == 1)
		copy(0, "standard input");
	else while (--argcnt) {
		arg = *++argvar;
		/* '-' means read from standard input */
		if (strcmp(name, "lp") == 0 && strcmp(arg, "-") == 0) {
			copy(0, "standard input");
			continue;
		}
		if ((f = test(arg)) < 0)
			continue;	/* file unreasonable */

		if (sflag && (cp = linked(arg)) != NULL) {
			(void) sprintf(buf, "%u %ul",
					(u_short)statb.st_dev, statb.st_ino);
			card('S', buf);
			if (format == 'p')
				card('T', title ? title : arg);
			for (i = 0; i < ncopies; i++)
				card(format, &dfname[inchar-2]);
			card('U', &dfname[inchar-2]);
			if (f)
				card('U', cp);
			card('N', arg);
			dfname[inchar]++;
			nact++;
			continue;
		}
		if (sflag)
			printf("%s: %s: not linked, copying instead\n", name, arg);
		if ((i = open(arg, O_RDONLY)) < 0) {
			printf("%s: cannot open %s\n", name, arg);
			continue;
		}
		copy(i, arg);
		(void) close(i);
		if (f && unlink(arg) < 0)
			printf("%s: %s: not removed\n", name, arg);
	}

	if (nact) {
		(void) close(tfd);
		tfname[inchar]--;
		/*
		 * Touch the control file to fix position in the queue.
		 */
		if ((tfd = open(tfname, O_RDWR)) >= 0) {
			char c;

			if (read(tfd, &c, 1) == 1 &&
			    lseek(tfd, 0L, L_SET) == 0 &&
			    write(tfd, &c, 1) != 1) {
				printf("%s: cannot touch %s\n", name, tfname);
				tfname[inchar]++;
				cleanup();
			}
			(void) close(tfd);
		}
		if (link(tfname, cfname) < 0) {
			printf("%s: cannot rename %s\n", name, cfname);
			tfname[inchar]++;
			cleanup();
		}
		(void)unlink(tfname);
		if (qflag)		/* just q things up */
			exit(0);
		if (!startdaemon(printer))
			printf("jobs queued, but cannot start daemon.\n");
		/* if this is lp, return the request id to the user */
		if (strcmp(name, "lp") == 0 && !silent)
			printf("request id is %d\n", jobno);
		exit(0);
	}
	cleanup();
	/* NOTREACHED */
}

static void
lprarg()
{
	register char *arg;
	int i;

	while (argcnt > 1 && argvar[1][0] == '-') {
		argcnt--;
		arg = *++argvar;
		switch (arg[1]) {

		case 'P':		/* specifiy printer name */
			if (arg[2])
				printer = &arg[2];
			else if (argcnt > 1) {
				argcnt--;
				printer = *++argvar;
			}
			break;

		case 'C':		/* classification spec */
			hdr++;
			if (arg[2])
				class = &arg[2];
			else if (argcnt > 1) {
				argcnt--;
				class = *++argvar;
			}
			break;

		case 'J':		/* job name */
			hdr++;
			if (arg[2])
				jobname = &arg[2];
			else if (argcnt > 1) {
				argcnt--;
				jobname = *++argvar;
			}
			break;

		case 'T':		/* pr's title line */
			if (arg[2])
				title = &arg[2];
			else if (argcnt > 1) {
				argcnt--;
				title = *++argvar;
			}
			break;

		case 'l':		/* literal output */
		case 'p':		/* print using ``pr'' */
		case 't':		/* print troff output (cat files) */
		case 'n':		/* print ditroff output */
		case 'd':		/* print tex output (dvi files) */
		case 'g':		/* print graph(1G) output */
		case 'c':		/* print cifplot output */
		case 'v':		/* print vplot output */
			format = arg[1];
			break;

		case 'f':		/* print fortran output */
			format = 'r';
			break;

		case '4':		/* troff fonts */
		case '3':
		case '2':
		case '1':
			if (argcnt > 1) {
				argcnt--;
				fonts[arg[1] - '1'] = *++argvar;
			}
			break;

		case 'w':		/* versatec page width */
			width = arg+2;
			break;

		case 'r':		/* remove file when done */
			rflag++;
			break;

		case 'm':		/* send mail when done */
			mailflg++;
			break;

		case 'h':		/* toggle want of header page */
			hdr = !hdr;
			break;

		case 's':		/* try to link files */
			sflag++;
			break;

		case 'q':		/* just q job */
			qflag++;
			break;

		case 'i':		/* indent output */
			iflag++;
			indent = arg[2] ? atoi(&arg[2]) : 8;
			break;

		case '#':		/* n copies */
			if (isdigit(arg[2])) {
				i = atoi(&arg[2]);
				if (i > 0)
					ncopies = i;
			}
		}
	}
	if (printer == NULL && (printer = getenv("PRINTER")) == NULL)
		printer = DEFLP;
	chkprinter(printer);
}

static void
lparg()
{
	char *arg;
	int copyflg=0;
	int i;

	while (argcnt > 1 && argvar[1][0] == '-') {
		argcnt--;
		arg = *++argvar;
		switch (arg[1]) {
			case 'c': /* copy files instead of creating links */
				copyflg++;
				break;
			case 'd':		/* specify destination */
				if (arg[2])
					printer = &arg[2];
				else if (argcnt > 1) {
					argcnt--;
					printer = *++argvar;
				}
				break;
			case 'm':
				mailflg++;
				break;
			case 'n':		/* n copies */
				if (isdigit(arg[2])) {
					i = atoi(&arg[2]);
					if (i > 0)
						ncopies = i;
				}
				break;
			case 'o':	/* printer specific options, no
					  way to pass this to the printer */
				break;
			case 's':
				silent++;
				break;
			case 't':
				hdr++;
				banner++;
				if (arg[2])
					bannerttl = &arg[2];
				else if (argcnt > 1) {
					argcnt--;
					bannerttl = *++argvar;
				}
				break;
			case 'w':
				wrtflg++;
				break;
			default:
				if (strcmp(arg, "-") == 0) {  /* backtrack */
					argcnt++;
					argvar--;
				}
				else
					if (!silent)
						fatal("lp: [-c] [-ddest] [-m] [-nnumber] [-ooption] [-s] -[ttitle] [-w] files");

		}
	}
	if (printer == NULL)
		if (((printer = getenv("LPDEST")) == NULL) &&
				 ((printer = getenv("PRINTER")) == NULL))
			printer = DEFLP;
	chkprinter(printer);

	if (copyflg == 0 && LP != NULL && *LP != '\0')
		sflag++;
}
/*
 * Create the file n and copy from file descriptor f.
 */
static void
copy(f, n)
	int f;
	char n[];
{
	register int fd, i, nr = 0, nc = 0, first = 1;
	char *tn;
	char buf[BUFSIZ];

	tn = malloc((u_int)(strlen(dfname) - (inchar - 2) + 1));
	if (tn == NULL)
		fatal("out of memory");
	strcpy(tn, &dfname[inchar - 2]); /* save current temp file name */
	fd = nfile(dfname);
	while ((i = read(f, buf, BUFSIZ)) > 0) {
		if (first) {
			first = 0;
			if (i >= 2 && buf[0] == '\100' && buf[1] == '\357' &&
			    format != 't') {
				printf(
			"%s: %s is a troff file, assuming '-t'\n", name, n);
				format = 't';
			}
		}
		if (write(fd, buf, i) != i) {
			printf("%s: %s: temp file write error\n", name, n);
			break;
		}
		nc += i;
		if (nc >= BUFSIZ) {
			nc -= BUFSIZ;
			nr++;
			if (MX > 0 && nr > MX) {
				printf("%s: %s: copy file is too large\n", name, n);
				break;
			}
		}
	}
	(void) close(fd);
	if (nc==0 && nr==0) {
		printf("%s: %s: empty input file\n", name, n);
		free(tn);
		return;
	}
	nact++;
	if (format == 'p')
		card('T', title ? title : n);
	for (i = 0; i < ncopies; i++)
		card(format, tn);
	card('U', tn);
	card('N', n);
	free(tn);
}

/*
 * Try and link the file to dfname. Return a pointer to the full
 * path name if successful.
 */
static char *
linked(file)
	register char *file;
{
	register char *cp;
	static char buf[MAXPATHLEN];

	if (*file != '/') {
		if (getwd(buf) == NULL)
			return NULL;
		while (file[0] == '.') {
			switch (file[1]) {
			case '/':
				file += 2;
				continue;
			case '.':
				if (file[2] == '/') {
					if ((cp = rindex(buf, '/')) != NULL)
						*cp = '\0';
					file += 3;
					continue;
				}
			}
			break;
		}
		strcat(buf, "/");
		strcat(buf, file);
		file = buf;
	}
	return (symlink(file, dfname) ? NULL : file);
}

/*
 * Put a line into the control file.
 */
static void
card(c, p2)
	register char c, *p2;
{
	char buf[BUFSIZ];
	register char *p1 = buf;
	register int len = 2;

	*p1++ = c;
	while ((c = *p2++) != '\0') {
		*p1++ = (c == '\n') ? ' ' : c;
		len++;
	}
	*p1++ = '\n';
	(void)write(tfd, buf, len);
}

/*
 * Create a new file in the spool directory.
 */
static int
nfile(n)
	char *n;
{
	register int f;
	int oldumask = umask(0);		/* should block signals */

        f = open(n, O_WRONLY|O_CREAT|O_EXCL ,FILMOD);

	(void) umask(oldumask);
	if (f < 0) {
		printf("%s: cannot create %s\n", name, n);
		cleanup();
	}
	if (fchown(f, userid, -1) < 0) {
		printf("%s: cannot chown %s\n", name, n);
		cleanup();
	}
	if (++n[inchar] > 'z') {
		if (++n[inchar-2] == 't') {
			printf("too many files - break up the job\n");
			cleanup();
		}
		n[inchar] = 'A';
	} else if (n[inchar] == '[')
		n[inchar] = 'a';
	return f;
}

/*
 * Cleanup after interrupts and errors.
 */
static void
cleanup()
{
	register int i;

	(void) signal(SIGHUP, SIG_IGN);
	(void) signal(SIGINT, SIG_IGN);
	(void) signal(SIGQUIT, SIG_IGN);
	(void) signal(SIGTERM, SIG_IGN);
	i = inchar;
	if (tfname)
		do
			(void) unlink(tfname);
		while (tfname[i]-- != 'A');
	if (cfname)
		do
			(void) unlink(cfname);
		while (cfname[i]-- != 'A');
	if (dfname)
		do {
			do
				(void) unlink(dfname);
			while (dfname[i]-- != 'A');
			dfname[i] = 'z';
		} while (dfname[i-2]-- != 'd');
	exit(1);
}

/*
 * Test to see if this is a printable file.
 * Return -1 if it is not, 0 if its printable, and 1 if
 * we should remove it after printing.
 */
static int
test(file)
	char *file;
{
	struct exec execb;
	register int fd;
	register char *cp;

	if (access(file, R_OK) < 0) {
		printf("%s: cannot access %s\n", name, file);
		return (-1);
	}
	if (stat(file, &statb) < 0) {
		printf("%s: cannot stat %s\n", name, file);
		return (-1);
	}
	if ((statb.st_mode & S_IFMT) == S_IFDIR) {
		printf("%s: %s is a directory\n", name, file);
		return (-1);
	}
	if (statb.st_size == 0) {
		printf("%s: %s is an empty file\n", name, file);
		return (-1);
 	}
	if ((fd = open(file, O_RDONLY)) < 0) {
		printf("%s: cannot open %s\n", name, file);
		return (-1);
	}
	if (read(fd, (char *)&execb, sizeof execb) == sizeof execb)
		switch (execb.a_magic) {
		case A_MAGIC1:
		case A_MAGIC2:
		case A_MAGIC3:
#ifdef A_MAGIC4
		case A_MAGIC4:
#endif
			printf("%s: %s is an executable program", name, file);
			goto error1;

		case ARMAG:
			printf("%s: %s is an archive file", name, file);
			goto error1;

		default:
			cp = (char *)&execb;
			if (cp[0] == '\100' && cp[1] == '\357' && format != 't') {
				printf(
			"%s: %s is a troff file, assuming '-t'\n", name, file);
				format = 't';
			}
		}
	(void) close(fd);
	if (rflag) {
		if ((cp = rindex(file, '/')) == NULL) {
			if (access(".", W_OK) == 0)
				return (1);
		} else {
			*cp = '\0';
			fd = access(file, W_OK);
			*cp = '/';
			if (fd == 0)
				return (1);
		}
		printf("%s: %s: is not removable by you\n", name, file);
	}
	return (0);

error1:
	printf(" and is unprintable\n");
	(void) close(fd);
	return (-1);
}

/*
 * itoa - integer to string conversion
 */

#define MAXDIGITS 11

static char *
itoa(i)
	register int i;
{
	static char b[MAXDIGITS + 1];

	(void)sprintf(b, "%d", i);
	return b;
}

/*
 * Perform lookup for printer name or abbreviation --
 */

char pbuf[BUFSIZ/2], *bp;

static void
chkprinter(s)
	char *s;
{
	int status;
	char buf[BUFSIZ];
	extern char *pgetstr();

	if ((status = pgetent(buf, s)) < 0)
		fatal("cannot open printer description file");
	else if (status == 0)
		fatal("%s: unknown printer", s);
	bp = pbuf;
	AG = pgetstr("ag", &bp);
	if ((SD = pgetstr("sd", &bp)) == NULL)
		SD = DEFSPOOL;
	if ((LO = pgetstr("lo", &bp)) == NULL)
		LO = DEFLOCK;
	RG = pgetstr("rg", &bp);
	LP = pgetstr("lp", &bp);
	if ((MX = pgetnum("mx")) < 0)
		MX = DEFMX;
	if ((MC = pgetnum("mc")) < 0)
		MC = DEFMAXCOPIES;
	if ((DU = pgetnum("du")) < 0)
		DU = DEFUID;
	SC = pgetflag("sc");
}

/*
 * Make the temp files.
 */
static int
mktemps()
{
	register int len, fd, n;
	register char *cp;
	char buf[BUFSIZ];
	int id;

	(void) sprintf(buf, "%s/.seq", SD);
	if ((fd = open(buf, O_RDWR|O_CREAT, 0661)) < 0) {
		printf("%s: cannot create %s\n", name, buf);
		exit(1);
	}
	if (flock(fd, LOCK_EX)) {
		printf("%s: cannot lock %s\n", name, buf);
		exit(1);
	}
	n = 0;
	if ((len = read(fd, buf, sizeof buf)) > 0) {
		for (cp = buf; len--; ) {
			if (!isdigit(*cp))
				break;
			n = n * 10 + (*cp++ - '0');
		}
	}
	len = strlen(SD) + strlen(host) + 8;
	tfname = mktempfile("tf", n, len);
	cfname = mktempfile("cf", n, len);
	dfname = mktempfile("df", n, len);
	inchar = strlen(SD) + 3;
	id = n;
	n = (n + 1) % 1000;
	(void)lseek(fd, 0L, L_SET);
	sprintf(buf, "%03d\n", n);
	(void)write(fd, buf, strlen(buf));
	(void)close(fd);	/* unlocks as well */
	return id;
}

/*
 * Make a temp file name.
 */
static char *
mktempfile(id, num, len)
	char	*id;
	int	num, len;
{
	register char *s;
	extern char *malloc();

	if ((s = malloc((u_int)len)) == NULL)
		fatal("out of memory");
	(void) sprintf(s, "%s/%sA%03d%s", SD, id, num, host);
	return s;
}

/*VARARGS*/
static void
fatal(va_alist)
	va_dcl
{
	va_list args;
	char *fmt;

	(void)printf("%s: ", name);
	va_start(args);
	fmt = va_arg(args, char *);
	(void)vprintf(fmt, args);
	va_end(args);
	putchar('\n');
	exit(1);
}
